File: apps/api/app/schemas/project.py
"""Pydantic schemas for Project endpoints."""
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.enums import ApprovalStatus, ProjectPriority, ProjectStatus


class ProjectCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    priority: ProjectPriority = ProjectPriority.MEDIUM
    due_date: date | None = None
    manager_id: uuid.UUID | None = None  # the manager who will approve (optional for admins)


class ProjectUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    priority: ProjectPriority | None = None
    due_date: date | None = None
    project_status: ProjectStatus | None = None
    manager_id: uuid.UUID | None = None


class ProjectRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    title: str
    description: str | None
    owner_id: uuid.UUID
    owner_name: str | None = None
    manager_id: uuid.UUID
    manager_name: str | None = None
    priority: ProjectPriority
    approval_status: ApprovalStatus
    project_status: ProjectStatus
    due_date: date | None
    approved_at: datetime | None
    rejected_reason: str | None
    progress_percentage: float = 0.0
    created_at: datetime
    updated_at: datetime


class ProjectDecision(BaseModel):
    decision: ApprovalStatus  # approved | rejected
    reason: str | None = None

File: apps/mobile/src/types/project.ts
export type ProjectPriority = 'low' | 'medium' | 'high' | 'critical';

export type ProjectApprovalStatus = 'pending' | 'approved' | 'rejected' | 'escalated' | 'cancelled' | 'needs_clarification';

export type ProjectStatus =
  | 'draft'
  | 'pending_approval'
  | 'approved'
  | 'active'
  | 'on_hold'
  | 'completed'
  | 'rejected'
  | 'archived';

export type ProjectHealth = 'on_track' | 'at_risk' | 'delayed' | 'completed' | 'planning' | 'unknown';

export interface ProjectApiRecord {
  id: string;
  title: string;
  description: string | null;
  owner_id: string;
  manager_id: string;
  priority: ProjectPriority;
  approval_status: ProjectApprovalStatus;
  project_status: ProjectStatus;
  due_date: string | null;
  approved_at: string | null;
  rejected_reason: string | null;
  progress_percentage: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreatePayload {
  title: string;
  description?: string | null;
  priority?: ProjectPriority;
  due_date?: string | null;
  manager_id?: string | null;
}

export interface ProjectDecisionPayload {
  decision: 'approved' | 'rejected';
  reason?: string | null;
}

export interface ProjectListParams {
  approval_status?: ProjectApprovalStatus;
  project_status?: ProjectStatus;
  owner_id?: string;
  manager_id?: string;
}

export interface ProjectTaskRecord {
  id: string;
  project_id: string;
  title: string;
  status: string;
  due_date: string | null;
  assigned_to_name?: string | null;
  priority: ProjectPriority;
}

export interface ProjectTaskSummary {
  total: number;
  completed: number;
  overdue: number;
  pending: number;
  inProgress: number;
}

export interface ProjectViewModel {
  id: string;
  name: string;
  description: string | null;
  status: ProjectStatus;
  approvalStatus: ProjectApprovalStatus;
  health: ProjectHealth;
  progress: number;
  priority: ProjectPriority;
  dueDate: string | null;
  createdAt: string;
  updatedAt: string;
  ownerId: string;
  managerId: string;
  ownerName?: string | null;
  managerName?: string | null;
  teamMembersCount: number;
  taskSummary: ProjectTaskSummary;
  canEdit: boolean;
  canApprove: boolean;
  canCreateTask: boolean;
  rejectedReason: string | null;
}
